File: festival-web/src/main/java/cn/bdqfork/web/route/message/resolver/ParameterResolverFactory.java
package cn.bdqfork.web.route.message.resolver;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * @author bdq
 * @since 2020/2/13
 */
public class ParameterResolverFactory {
    private List<AbstractParameterResolver> parameterResolvers = new ArrayList<>();

    public ParameterResolverFactory() {
        registerResolver(new ContextParameterResolver());
        registerResolver(new PrimitiveParameterResolver());
        registerResolver(new ObjectParameterResolver());
        registerResolver(new JsonBodyParameterResolver());
        registerResolver(new DateParameterResolver());
    }

    public void registerResolver(AbstractParameterResolver parameterResolver) {
        parameterResolvers.add(parameterResolver);
    }

    public void registerResolver(Collection<AbstractParameterResolver> parameterResolvers) {
        this.parameterResolvers.addAll(parameterResolvers);
    }

    public ParameterResolver createResolverChain() {
        if (parameterResolvers != null && parameterResolvers.size() > 0) {
            for (int i = 0; i < parameterResolvers.size() - 1; i++) {
                AbstractParameterResolver resolver = parameterResolvers.get(i);
                AbstractParameterResolver next = parameterResolvers.get(i + 1);
                resolver.setNext(next);
            }
            return parameterResolvers.get(0);
        }
        return null;
    }

}

File: festival-kotlin-web/src/main/kotlin/cn/bdqfork/kotlin/web/route/message/resolver/ParameterResolverFactory.kt
package cn.bdqfork.web.route.message.resolver

import java.util.*

/**
 * @author bdq
 * @since 2020/2/13
 */
class ParameterResolverFactory {
    private val parameterResolvers: MutableList<AbstractParameterResolver> = ArrayList()
    private fun registerResolver(parameterResolver: AbstractParameterResolver) {
        parameterResolvers.add(parameterResolver)
    }

    fun registerResolver(parameterResolvers: Collection<AbstractParameterResolver>) {
        this.parameterResolvers.addAll(parameterResolvers)
    }

    fun createResolverChain(): ParameterResolver {
        for (i in 0 until parameterResolvers.size - 1) {
            val resolver = parameterResolvers[i]
            val next = parameterResolvers[i + 1]
            resolver.setNext(next)
        }
        return parameterResolvers[0]
    }

    init {
        registerResolver(ContextParameterResolver())
        registerResolver(PrimitiveParameterResolver())
        registerResolver(JsonBodyParameterResolver())
        registerResolver(DateParameterResolver())
    }
}